src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -202,42 +207,60 @@ private void EventLoop()
             }
         }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void ScheduleToProcessEvents()
+        private void UpdateEventQueueProcessingStage(bool isEventQueueEmpty)
         {
-            // Schedule a thread pool work item to process events. Only one work item is scheduled at any given time to avoid
-            // over-parallelization. When the work item begins running, this field is reset to 0, allowing for another work item
-            // to be scheduled for parallelizing processing of events.
-            if (Interlocked.CompareExchange(ref _eventQueueProcessingRequested, 1, 0) == 0)
+            if (!isEventQueueEmpty)
             {
-                ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+                _eventQueueProcessingStage = (int)EventQueueProcessingStage.Scheduled;
             }
+            else
+            {
+                int stageBeforeUpdate =
+                    Interlocked.CompareExchange(
+                        ref _eventQueueProcessingStage,
+                        (int)EventQueueProcessingStage.NotScheduled,
+                        (int)EventQueueProcessingStage.Determining);
+                Debug.Assert(stageBeforeUpdate != (int)EventQueueProcessingStage.NotScheduled);
+                if (stageBeforeUpdate == (int)EventQueueProcessingStage.Determining)
+                {
+                    return;
+                }
+            }
+
+            ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
         }
 
         void IThreadPoolWorkItem.Execute()
         {
-            // Indicate that a work item is no longer scheduled to process events. The change needs to be visible to enqueuer
-            // threads (only for EventLoop() currently) before an event is attempted to be dequeued. In particular, if an
-            // enqueuer queues an event and does not schedule a work item because it is already scheduled, and this thread is
-            // the last thread processing events, it must see the event queued by the enqueuer.
-            Interlocked.Exchange(ref _eventQueueProcessingRequested, 0);
-
             ConcurrentQueue<SocketIOEvent> eventQueue = _eventQueue;
-            if (!eventQueue.TryDequeue(out SocketIOEvent ev))
+            SocketIOEvent ev;
+            while (true)
             {
-                return;
-            }
+                Debug.Assert(_eventQueueProcessingStage == (int)EventQueueProcessingStage.Scheduled);
+                _eventQueueProcessingStage = (int)EventQueueProcessingStage.Determining;
+                Interlocked.MemoryBarrier();
```

Comment:
A memory barrier needs to be added to a similar place for the other two cases as well

Reply:
Right, I'll add those